typescript-loader/loader.js

```diff
@@ -0,0 +1,46 @@
+// 💡 It's probably a good idea to put this loader near the end of the chain.
+
+import path from 'node:path';
+
+import { transform } from 'esbuild';
+
+
+const tsExts = new Set([
+  '.tsx',
+  '.ts',
+  '.mts',
+  '.cts',
+]);
+
+export async function resolve(specifier, context, nextResolve) {
+  const ext = path.extname(specifier);
+
+  const resolution = nextResolve(specifier); // No need to await nextResolve (yet)
+
+  if (!tsExts.has(ext)) { return resolution; } // File is not ts, so step aside
+
+  return {
+    format: 'typescript', // Provide a signal to `load`
+    shortCircuit: true,
+    url: (await resolution).url,
```

Comment:
How is `resolution` achieved? Node’s built-in `resolve` will resolve a `.ts` file? TypeScript has all these additional rules like `"paths"` in `tsconfig.json`; don’t we need to involve some library like `tsc` or `esbuild` to handle TypeScript resolution?

Reply:
Per comments above, pretty sure the goal was to avoid fancy stuff to keep this loader simple, with the understanding that third-party loader implementations offer a more complete solution.

Also, stuff like `"paths"` is meant to inform the typechecker about the runtime's resolution behavior, not to modify the runtime's resolution behavior.  Depends on who you ask of course, plenty of folks believe that runtimes should implement `"paths"`.  But the whole idea with TS's new NodeNext mode is that you embrace the runtime's native resolution behaviors, and TS's checker will resolve the same way when doing semantic analysis.

Reply:
What Andrew said 🙂

If not doing anything typescript-config specific, it works (run `npm test` for proof). If you want more integrated typescript functionality, we decided "use ts-node".

Reply:
Leaving aside `"paths"` etc, how does this work at all? I wouldn’t expect `resolution` to have worked since the specifier was a `.ts` file and Node shouldn’t know how to resolve those (?)

Reply:
I think you're conflating resolve's validation and load's validation: resolve only cares that it exists. Resolve does NOT care what it is (load does).